### PR TITLE
feat: add controller builder options and funcs

### PR DESCRIPTION
**Pull request by kinsolee**
# Changes
 add controller builder options and funcs

# Submitter Checklist

As the author of this PR, please check off the items in this checklist:

- [ ] [`spec` PR link](https://github.com/katanomi/spec) included
- [ ] Follows the [commit message standard](https://github.com/katanomi/spec/blob/main/CONTRIBUTING.md#commits)
- [ ] Meets the [contributing guidelines](https://github.com/katanomi/spec/blob/main/CONTRIBUTING.md) (including
  functionality, content, code)
- [ ] Test cases with documentation and functionality works as expected using current and related github repos (MUST deploy and check)
- [ ] Release notes block below has been filled in or deleted (only if no user facing changes)

# Release Notes

### CODE REVIEW

File: controllers/builderfuncs.go [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+import (
+	"sigs.k8s.io/controller-runtime/pkg/builder"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+)
+
+// ControllerBuilderFunc is a function that can be used to configure the controller builder
+type ControllerBuilderFunc func(*builder.Builder) *builder.Builder
+
+// ApplyControllerBuilderFuncs applies the given options to the controller builder
+func ApplyControllerBuilderFuncs(builder *builder.Builder, funcs ...ControllerBuilderFunc) *builder.Builder {
```
Comment (kinsolee) [inline controllers/builderfuncs.go:28]: 这里是对controller builder做配置，为了避免跟builder options混淆，改成了builderFunc

File: controllers/builderoptions.go [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+
+// DefaultMaxConcurrentReconciles is the default number of max concurrent reconciles
+const DefaultMaxConcurrentReconciles = 10
+
+// BuilderOptionFunc is a function that can be used to configure the controller builder options
+type BuilderOptionFunc func(options controller.Options) controller.Options
+
+// BuilderOptions returns a functional set of options with conservative
+// defaults.
+func BuilderOptions(opts ...BuilderOptionFunc) controller.Options {
```
Comment (kinsolee) [inline controllers/builderoptions.go:32]: 这里业务端一般调用这个方法，传入额外的OptionFunc覆盖默认值

File: controllers/builderfuncs.go (hunk 1 above)
Comment (nanjingfm) [reply controllers/builderfuncs.go:28]: 下面的叫builderOption,这个叫controllerOption?

Comment (kinsolee) [reply controllers/builderfuncs.go:28]: 经过讨论上面改成 ControllerBuilderOption

**Review by nanjingfm — APPROVED** [review]

**Pull request merged**